### PR TITLE
Row-level partitioning support in YSQL

### PR DESCRIPTION
**Pull request by deeps1991**

### CODE REVIEW

File: src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+SET max_parallel_workers_per_gather TO 0;
+--
+-- Tests for list partitioned tables.
+--
+CREATE TABLE pagg_tab (a int, b int, c text, d int) PARTITION BY LIST(c);
+CREATE TABLE pagg_tab_p1 PARTITION OF pagg_tab FOR VALUES IN ('0000', '0001', '0002', '0003');
+CREATE TABLE pagg_tab_p2 PARTITION OF pagg_tab FOR VALUES IN ('0004', '0005', '0006', '0007');
+CREATE TABLE pagg_tab_p3 PARTITION OF pagg_tab FOR VALUES IN ('0008', '0009', '0010', '0011');
+INSERT INTO pagg_tab SELECT i % 20, i % 30, to_char(i % 12, 'FM0000'), i % 30 FROM generate_series(0, 2999) i;
+-- When GROUP BY clause matches; full aggregation is performed for each partition.
```
Comment (m-iancu) [inline src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out:19]: I notice the EXPLAIN tests are removed from the original. Did they not work or were they unreliable?
I think it's good to have them here -- if they don't yet work we can comment them out and add a comment explaining what the limitation is

File: src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+CREATE TABLE pagg_tab_ml_p2 PARTITION OF pagg_tab_ml FOR VALUES FROM (10) TO (20) PARTITION BY LIST (c);
+CREATE TABLE pagg_tab_ml_p2_s1 PARTITION OF pagg_tab_ml_p2 FOR VALUES IN ('0000', '0001');
+CREATE TABLE pagg_tab_ml_p2_s2 PARTITION OF pagg_tab_ml_p2 FOR VALUES IN ('0002', '0003');
+-- This level of partitioning has different column positions than the parent
+CREATE TABLE pagg_tab_ml_p3 PARTITION OF pagg_tab_ml FOR VALUES FROM (20) TO (30) PARTITION BY RANGE (b);
+CREATE TABLE pagg_tab_ml_p3_s1 PARTITION OF pagg_tab_ml_p3 FOR VALUES FROM (0) TO (5);
+CREATE TABLE pagg_tab_ml_p3_s2 PARTITION OF pagg_tab_ml_p3 FOR VALUES FROM (5) TO (10);
+INSERT INTO pagg_tab_ml SELECT i % 30, i % 10, to_char(i % 4, 'FM0000') FROM generate_series(0, 29999) i;
+-- For Parallel Append
+SET max_parallel_workers_per_gather TO 2;
```
Comment (m-iancu) [inline src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out:309]: Do we actually support parallel execution for this case? (The explain plan should have a `Parallel Append` node if we do.

File: src/postgres/src/test/regress/expected/yb_pg_partition_join.out [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+ 100 | 0 | 0000 |     |      |     |     |      
+ 150 | 0 | 0002 | 150 | 0002 |   0 | 150 |   150
+ 200 | 0 | 0000 |     |      |     |     |      
+ 250 | 0 | 0002 |     |      |     |     |      
+ 300 | 0 | 0000 | 300 | 0000 |   0 | 300 |   300
+ 350 | 0 | 0002 |     |      |     |     |      
+ 400 | 0 | 0000 |     |      |     |     |      
+ 450 | 0 | 0002 | 450 | 0002 |   0 | 450 |   450
+ 500 | 0 | 0000 |     |      |     |     |      
+ 550 | 0 | 0002 |     |      |     |     |
```
Comment (m-iancu) [inline src/postgres/src/test/regress/expected/yb_pg_partition_join.out:535]: ditto about explain.

File: src/postgres/src/test/regress/sql/yb_pg_triggers.sql [hunk 4]
```diff
@@ -1316,12 +1314,17 @@ create table trigpart1 partition of trigpart for values from (0) to (1000);
...
 create table trigpart3 (like trigpart);
-alter table trigpart attach partition trigpart3 for values from (2000) to (3000);
 select tgrelid::regclass, tgname, tgfoid::regproc from pg_trigger
   where tgrelid::regclass::text like 'trigpart%' order by tgrelid::regclass::text;
+-- Disabling the below tests temporarily because each run of the test can
+-- throw one of two errors. The error message displayed contains the first dependency
+-- resulting from the scan of pg_depend preventing the delete of the trigger. The trigger
+-- depends on the partition table itself and on the trigger created on the partitioned
+-- table, and either of the two error messages are possible.
+/*
```
Comment (m-iancu) [inline src/postgres/src/test/regress/sql/yb_pg_triggers.sql:1324]: For this cases, instead of disabling them, we temporarily change verbosity level for these commands to omit the detail lines.
For instance:
```
\set verbosity terse
DROP .. 
\set verbosity default
```
Alternatively, you can use the native SQL way of doing `SET client_min_messages TO 'WARNING'` and then `RESET client_min_messages`.
The latter is probably cleaner if it works because it's just native SQL (rather than a `ysqlsh` command) but (at least) one of them typically works to stabilize test output in these cases.

File: src/postgres/src/backend/commands/tablecmds.c [hunk 5]
```diff
@@ -796,9 +796,6 @@ DefineRelation(CreateStmt *stmt, char relkind, Oid ownerId,
 										  InvalidOid,
 										  typaddress);
 
-	/* Store inheritance information for new rel. */
```
Comment (m-iancu) [inline src/postgres/src/backend/commands/tablecmds.c:799]: Why was this removed?

**Comment by deeps1991** [comment]
I seem to have messed something up while rebasing and pushing, will submit a new pull request. Sorry about the confusion!

File: src/postgres/src/backend/utils/misc/pg_yb_utils.c [hunk 6]
```diff
@@ -82,7 +82,9 @@ CheckIsYBSupportedRelationByKind(char relkind)
 {
 	if (!(relkind == RELKIND_RELATION || relkind == RELKIND_INDEX ||
 		  relkind == RELKIND_VIEW || relkind == RELKIND_SEQUENCE ||
-		  relkind == RELKIND_COMPOSITE_TYPE))
+		  relkind == RELKIND_COMPOSITE_TYPE || relkind == RELKIND_PARTITIONED_TABLE ||
+		  relkind == RELKIND_PARTITIONED_INDEX))
+
```
[the remark this answers is not in the document]
Comment (deeps1991) [reply src/postgres/src/backend/utils/misc/pg_yb_utils.c]: This somehow doesn't show up on my git diff here, but fixed now. Thanks for pointing this out!

File: src/postgres/src/backend/catalog/partition.c [hunk 7]
```diff
@@ -240,7 +240,10 @@ has_partition_attrs(Relation rel, Bitmapset *attnums, bool *used_in_expr)
 			Bitmapset  *expr_attrs = NULL;
 
 			/* Find all attributes referenced */
-			pull_varattnos(expr, 1, &expr_attrs);
+			// pull_varattnos_min_attr offsets attributes by min_attr + 1. Therefore pass
+			// YBGetFirstLowInvalidAttributeNumber + 1.
```
[the remark this answers is not in the document]
Comment (deeps1991) [reply src/postgres/src/backend/catalog/partition.c]: Fixed

File: src/postgres/src/backend/catalog/partition.c [hunk 8]
```diff
@@ -225,8 +225,8 @@ has_partition_attrs(Relation rel, Bitmapset *attnums, bool *used_in_expr)
 
 		if (partattno != 0)
 		{
-			if (bms_is_member(partattno - FirstLowInvalidHeapAttributeNumber,
-							  attnums))
+			if (bms_is_member(partattno - YBGetFirstLowInvalidAttributeNumber(rel),
+					  	attnums))
```
[the remark this answers is not in the document]
Comment (deeps1991) [reply src/postgres/src/backend/catalog/partition.c]: Fixed

File: src/postgres/src/backend/executor/nodeModifyTable.c [hunk 9]
```diff
@@ -1133,6 +1127,10 @@ ExecUpdate(ModifyTableState *mtstate,
 		if (resultRelationDesc->rd_att->constr)
 			ExecConstraints(resultRelInfo, slot, estate, mtstate);
 
+		/* 
+		 * Verify that the update does not violate partition constraints.
+         */
```
[the remark this answers is not in the document]
Comment (deeps1991) [reply src/postgres/src/backend/executor/nodeModifyTable.c]: Fixed

File: src/postgres/src/backend/optimizer/prep/prepunion.c [hunk 10]
```diff
@@ -115,8 +116,9 @@ static void make_inh_translation_list(Relation oldrelation,
 						  Relation newrelation,
 						  Index newvarno,
 						  List **translated_vars);
-static Bitmapset *translate_col_privs(const Bitmapset *parent_privs,
-					List *translated_vars);
+static Bitmapset *translate_col_privs(Relation parentrel,
+					const Bitmapset *parent_privs,
+                    List *translated_vars);
```
[the remark this answers is not in the document]
Comment (deeps1991) [reply src/postgres/src/backend/optimizer/prep/prepunion.c]: Fixes

File: src/postgres/src/test/regress/expected/yb_pg_foreign_key.out [hunk 11]
```diff
@@ -1692,14 +1760,49 @@ ALTER TABLE fk_partitioned_fk DETACH PARTITION fk_partitioned_fk_4;
...
+    "fk_partitioned_fk_4_a_fkey" FOREIGN KEY (a, b) REFERENCES fk_notpartitioned_pk(a, b) ON UPDATE CASCADE ON DELETE CASCADE
 
 CREATE TABLE fk_partitioned_fk_5 (a int, b int,
-	FOREIGN KEY (a, b) REFERENCES fk_notpartitioned_pk(a, b) ON UPDATE CASCADE ON DELETE CASCADE DEFERRABLE,
-	FOREIGN KEY (a, b) REFERENCES fk_notpartitioned_pk(a, b) MATCH FULL ON UPDATE CASCADE ON DELETE CASCADE)
-  PARTITION BY RANGE (a);
+FOREIGN KEY (a, b) REFERENCES fk_notpartitioned_pk(a, b) ON UPDATE CASCADE ON DELETE CASCADE DEFERRABLE,
+FOREIGN KEY (a, b) REFERENCES fk_notpartitioned_pk(a, b) MATCH FULL ON UPDATE CASCADE ON DELETE CASCADE)
+PARTITION BY RANGE (a);
+ERROR:  DEFERRABLE constraint not supported yet
```
[the remark this answers is not in the document]
Comment (deeps1991) [reply src/postgres/src/test/regress/expected/yb_pg_foreign_key.out]: Thanks for pointing this out! Rebased and fixed these now

File: src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out (hunk 1 above)
Comment (deeps1991) [reply src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out:24]: Thanks for reminding me about these! Sorry that I missed them out. A good number of them don't work. I now added all of them back, and have left the ones that don't work commented out. If you could create another ticket and assign it to me, I will take care of those as part of a separate task after this one is pushed.
While exploring this I noticed I missed out partition_prune tests. Added those as well.

File: src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out (hunk 2 above)
Comment (deeps1991) [reply src/postgres/src/test/regress/expected/yb_pg_partition_aggregate.out:736]: The expected output seems to expect a Parallel Append, but I didn't seem to see that happen, have commented them out. I have to explore it further, which I will do as part of a separate task :)

File: src/postgres/src/test/regress/sql/yb_pg_triggers.sql (hunk 4 above)
Comment (deeps1991) [reply src/postgres/src/test/regress/sql/yb_pg_triggers.sql]: The problem here was that was differing error messages... And there doesn't seem to be a way to silence Error messages as far as I could google :( Is there anything else I can do?

File: src/postgres/src/backend/commands/tablecmds.c (hunk 5 above)
Comment (deeps1991) [reply src/postgres/src/backend/commands/tablecmds.c:799]: I have explained this in detail in the commit message. I think this is a mistaken merge conflict artifact, as I couldn't see this line in Postgres 11.2 code.
